lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,153 @@
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  *
+  * @details The class handles symbolic expressions over values read from hardware graphics registers, materialized as
+  *          llvm instructions. The code maintains a map from a register ID to an llvm value with the current symbolic
+  *          expression describing the updated state of the register.
+  *
+  * @sa llpc/docs/DdnGraphicsRegisterHandler.md
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce
+// optimized llvm instruction sequences for these states.
+// - pValue : Cache the value which is already being fetched.
+// - isModified : True, if the register is being settled with new value.
+struct BitsState
+{
+    Value* pValue     = nullptr;
+    bool   isModified = false;
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers.
+// Note: Don't use GfxRegHelper directly, please implement your own register helper class, such as SqImgSampRegHelper
+class GfxRegHelper : public GfxRegHandlerBase
+{
+protected:
+    GfxRegHelper(Builder* pBuilder, Value* pRegister);
+
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
+
+    // Set register
+    void SetRegCommon(uint32_t id, Value* pVal)
+    {
+        SetBits(m_bitsInfo[id], pVal);
+        // It is assumed the register is being modified
+        m_bitsState[id].isModified = 1;
+    }
+
+    // Get combined data from two seperate DWORDs
+    Value* GetRegCombine(uint32_t idLo, uint32_t idHi);
+
+    // Set data into two seperate DWORDs
+    void SetRegCombine(uint32_t idLo, uint32_t idHi, Value* pReg);
+
+protected:
+    Value*                       m_pInt32One     = nullptr;
+    GfxIpVersion*                m_pGfxIpVersion = nullptr;
+    const BitsInfo*              m_bitsInfo      = nullptr;
+    std::unique_ptr<BitsState[]> m_bitsState     = nullptr;
```

Comment:
Can we have a comment that says that this is a map from register ID (?) to its current state? And because the number of IDs is small, it's implemented with an array.

Reply:
Comment on the top of the class now.